src/librbd/Lock.cc

```diff
@@ -0,0 +1,620 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
+// vim: ts=8 sw=2 smarttab
+
+#include "librbd/Lock.h"
+#include "librbd/managed_lock/Policy.h"
+#include "librbd/managed_lock/AcquireRequest.h"
+#include "librbd/managed_lock/ReleaseRequest.h"
+#include "librbd/managed_lock/ReacquireRequest.h"
+#include "librbd/managed_lock/NoReleasePolicy.h"
+#include "cls/lock/cls_lock_client.h"
+#include "common/dout.h"
+#include "common/errno.h"
+#include "common/WorkQueue.h"
+#include "librbd/Utils.h"
+#include <sstream>
+
+#define dout_subsys ceph_subsys_rbd
+#undef dout_prefix
+#define dout_prefix *_dout << "librbd::Lock: "
+
+namespace librbd {
+
+using namespace managed_lock;
+using std::string;
+using util::detail::C_AsyncCallback;
+
+namespace {
+
+template <typename R>
+struct C_SendRequest : public Context {
+  R* request;
+  explicit C_SendRequest(R* request) : request(request) {
+  }
+  virtual void finish(int r) override {
+    request->send();
+  }
+};
+
+class ThreadPoolSingleton : public ThreadPool {
```

Comment:
The reason of why I used a new thread pool and consequently a new 'ContextWQ' was because the 'op_work_queue' from the 'ImageCTx' class is being mocked in the unit tests and I didn't want to make the 'Lock' class to depend on 'ImageCtx' as a template parameter of 'Lock'.